### ISSUE TITLE
icingadb mysql FATA[0001] Error 1064

### ISSUE THREAD

**Issue by mb-77** [Icinga/icingadb#209]
I get the follow error after a while icingadb is running:

INFO[0000] Connecting to Redis
INFO[0000] Connecting to MySQL
INFO[0000] Redis connection established
INFO[0000] Starting heartbeat listener
INFO[0000] SQL connection established
INFO[0000] Got initial environment.                      UUID=04bafde1-3d3a-4ee0-9a90-ac17affacdf3 context=HA environment=da39a3ee5e6b4b0d3255bfef95601890afd80709
INFO[0000] Taking over.                                  UUID=04bafde1-3d3a-4ee0-9a90-ac17affacdf3 context=HA environment=da39a3ee5e6b4b0d3255bfef95601890afd80709
INFO[0001] Inserted 2 customvar_flats in 1.856756ms      action=insert benchmark=1.856756ms count=2 type=customvar_flat
**FATA[0001] Error 1064: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '' at line 1**

I've try to drop and re-create the DB and its schema, but the problem persist

**Comment by Obihoernchen** [comment]
I experience the same issue.

**Comment by ziirish** [comment]
Same error.

```
Icinga DB version: v1.0.0-rc1
icinga2 - The Icinga 2 network monitoring daemon (version: r2.12.3-1)

ii  icingadb                             1.0.0~rc1-2.buster           amd64        database connector for Icinga 2
ii  icingadb-redis                       1.0.0~rc1-2.buster           amd64        Redis service for Icinga DB
ii  mariadb-common                   1:10.4.17+maria~buster      all          MariaDB database common files (e.g. /etc/mysql/conf.d/mariadb.cnf)
```

**Comment by ziirish** [comment]
With the debug logs:

``` 
Jan 07 10:05:48 master-eu-1-vision icingadb[27092]: time="2021-01-07T10:05:48Z" level=debug msg="notes_url: Insert execution worker started"
Jan 07 10:05:48 master-eu-1-vision icingadb[27092]: time="2021-01-07T10:05:48Z" level=debug msg="notes_url: Delete execution worker started"
Jan 07 10:05:48 master-eu-1-vision icingadb[27092]: time="2021-01-07T10:05:48Z" level=debug msg="notes_url: Update comparison worker started"
Jan 07 10:05:48 master-eu-1-vision icingadb[27092]: time="2021-01-07T10:05:48Z" level=debug msg="notes_url: Update preparation worker started"
Jan 07 10:05:48 master-eu-1-vision icingadb[27092]: time="2021-01-07T10:05:48Z" level=debug msg="notes_url: Update execution worker started"
Jan 07 10:05:48 master-eu-1-vision icingadb[27092]: time="2021-01-07T10:05:48Z" level=debug msg="Finished Exec" affected_rows=1 args="{[[204 109 31 196 78 36 67 35 172 192 20 24 127 159 61 33] [218 57 163 238 94 107 75 13 50 85 191 239 149 96 24 144 175 216 7 9] [169 170 75 180 236 120 75 57 38 117 137 135 34 159 176 183 197 113 70 152] service [62 52 10 159 153 112 187 0 246 74 90 73 71 238 118 152 148 8 62 33] [70 228 33 138 236 142 43 255 139 118 74 57 1 154 179 29 118 105 214 91] [4 30 193 145 26 84 79 194 172 22 53 249 183 127 212 104] <nil> <nil> <nil> <nil> notification 1609937728485]}" benchmark=1.481016ms context=sql query="REPLACE INTO history (id, environment_id, endpoint_id, object_type, host_id, service_id, notification_history_id,state_history_id, downtime_history_id, comment_history_id, flapping_history_id, event_type, event_time)VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?)"
Jan 07 10:05:48 master-eu-1-vision icingadb[27092]: time="2021-01-07T10:05:48Z" level=info msg="Inserted 2 customvar_flats in 9.295394ms" action=insert benchmark=9.295394ms count=2 type=customvar_flat
Jan 07 10:05:48 master-eu-1-vision icingadb[27092]: time="2021-01-07T10:05:48Z" level=fatal msg="Error 1064: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '' at line 1"
Jan 07 10:05:48 master-eu-1-vision systemd[1]: icingadb.service: Main process exited, code=exited, status=1/FAILURE
Jan 07 10:05:48 master-eu-1-vision systemd[1]: icingadb.service: Failed with result 'exit-code'.
```

**Comment by Al2Klimov** [comment]
Hello all!

Please could any of you reproduce the error while running `tshark -i any -f 'port 3306' -b duration:30 -w 209.pcap` on the Icinga DB host, wait 30s, interrupt tshark and upload the generated files [here](https://nextcloud.icinga.com/index.php/s/azsMrzCPtNC7jcc)?

Best,
AK

**Comment by mocdaniel** [comment]
Hello, I could reproduce the described bug and uploaded 3 generated files to the link you gave us, @Al2Klimov 

Interestingly, we only face this issue on one of two identical setups.

**Comment by Al2Klimov** [comment]
Interesting...

![Bildschirmfoto 2021-01-28 um 17 03 43](https://user-images.githubusercontent.com/6053266/106165197-22f8a100-618b-11eb-9ad5-42379a00ee29.png)

@N-o-X Any logical explanation for this?

**Comment by mocdaniel** [comment]
@N-o-X @Al2Klimov another thing that came to mind and which I could verify just now: Both of our machines use CentOS7, but only on the machine which can reproduce this bug has SELinux enabled. Could this be a SELinux problem?

**Comment by Al2Klimov** [comment]
https://github.com/Icinga/icingadb/issues/209#issuecomment-769191274 ? Never ever.

@dgoetz Feel free to correct me.

**Comment by ziirish** [comment]
On my side it happened on Debian 10 "vanilla", so without any selinux nor apparmor.

**Comment by julianbrost** [comment]
The only affected version reported was 1.0.0-rc1. I believe this was inadvertently fixed in ee27e9809b9dd425935b7b7aa9d0643fd8375115.

In 1.0.0-rc1, `DBWrapper.SqlBulkInsert()` returns instantly if 0 rows are to be inserted: https://github.com/Icinga/icingadb/blob/f6d915659bbadf24aff78f48cbd57369c9d0ee41/connection/mysql.go#L689-L691

However, right after that, on each row `GetFinalRows()` is called and the results are concatenated to replace the current slice of rows: https://github.com/Icinga/icingadb/blob/f6d915659bbadf24aff78f48cbd57369c9d0ee41/connection/mysql.go#L693-L699

So if all these rows return an empty slice in `GetFinalRows()`, this can lead to the error in this issue. And indeed, for `customvar_flat`, it looks like this can indeed return an empty slice for empty arrays or maps (however I wasn't able to reproduce this in a very quick test): https://github.com/Icinga/icingadb/blob/f6d915659bbadf24aff78f48cbd57369c9d0ee41/configobject/objecttypes/customvar/customvarflat/customvarflat.go#L68-L81

ee27e9809b9dd425935b7b7aa9d0643fd8375115 changed how rows are processed and they are now put through a new `ChunkRows()` function: https://github.com/Icinga/icingadb/blob/ee27e9809b9dd425935b7b7aa9d0643fd8375115/connection/mysql.go#L708

So unless this would return an empty chunk, this should no longer occur. @Al2Klimov just created #255 to additionally verify that this doesn't happen.

@mb-77 @Obihoernchen @ziirish Can you please try any of the recent snapshot builds and check if the error still occurs?
